src/AutoRest.CSharp/Common/Output/Models/Types/ModelFactoryTypeProvider.cs

```diff
@@ -31,30 +31,19 @@ internal sealed class ModelFactoryTypeProvider : TypeProvider
         protected override string DefaultName { get; }
         protected override string DefaultAccessibility { get; }
 
-        // TODO: remove this intermediate state once we generate it before output types
         private IReadOnlyList<Method>? _methods;
+        public IReadOnlyList<Method> Methods => _methods ??= BuildMethods().ToArray();
 
-        // This method should only be called from OutputMethods as intermediate state.
-        private IReadOnlyList<Method> ShouldNotBeUsedForOutput([CallerMemberName] string caller = "")
```

Comment:
This private method can be removed because of the change of `SignatureType`, which introduced additional properties to the base `TypeProvider`, I explained in the other comment.